pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java

```diff
@@ -60,6 +59,13 @@ public class PersistentAcknowledgmentsGroupingTracker implements Acknowledgments
      * When reaching the max group size, an ack command is sent out immediately.
      */
     private static final int MAX_ACK_GROUP_SIZE = 1000;
+    private static final FastThreadLocal<LastCumulativeAck> LOCAL_LAST_CUMULATIVE_ACK_TO_FLUSH =
```

Comment:
If use FastThreadLocal, multiple `PersistentAcknowledgmentsGroupingTrackers` will affect each other `LastCumulativeAck `

Reply:
Good catch!

Reply:
I think not. Because this field is just a temporary cache, it's only used in the following code block.

```java
            final LastCumulativeAck cumulativeAck = LOCAL_LAST_CUMULATIVE_ACK_TO_FLUSH.get();
            lastCumulativeAck.moveOwnershipTo(cumulativeAck);
            /* ... */
            if (!lastCumulativeAck.restoreOwnershipIfEmpty(cumulativeAck)) {
                // `lastCumulativeAck` has been updated so that restoreOwnershipIfEmpty failed
                cumulativeAck.recycle();
            }
            cumulativeAckFlushRequired = false;
```

The code above has no side effect on the `lastCumulativeAck` field if it's executed atomically.

Assuming there are two trackers, let's say `T1`, `T2`. The actual `lastCumulativeAck` objects are `T1.lastCumulativeAck` and `T2.lastCumulativeAck`.

If `T1.flushAsync()` and `T2.flushAsync()` are called in different threads, `LOCAL_LAST_CUMULATIVE_ACK_TO_FLUSH.get()` will return two different `LastCumulativeAck` objects and they won't affect each other.

If `T1.flushAsync()` and `T2.flushAsync()` are called in the same thread, these two calls cannot be executed concurrently. Each time `flushAsync()` is done, the `lastCumulativeAck` field of each tracker won't change unless it's updated in other methods.

Reply:
`LOCAL_LAST_CUMULATIVE_ACK_TO_FLUSH` is only used to store the value of `lastCumulativeAck` before sending the ACK so that we can avoid synchronizing the whole code block. After ACK is sent, the `lastCumulativeAck` will restore from the temporary value.